torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -662,16 +718,20 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
         local_used_maps_reduced_ = true;
       }
 
-      auto bucket_view =
-          replica.contents.narrow(0, offset, length).view(variable.sizes());
       auto& grad = variable.grad();
 
       // If a parameter is globally unused, we keep its grad untouched.
       if (!global_unused) {
         if (!grad.defined()) {
```

Comment:
How can that grad be undefined? 

Also to handle this easily, you could consider calling into `accumulateGrad` directly. (just like the distributed autograd engine does). With expected refs low enough that it never steals the given gradient.

Reply:
**deep breath** It can be undefined if the gradient was None at the beginning of the iteration, and this process didn't use this param, therefore didn't produce a gradient for it.  If DDP was constructed with `find_unused_parameters=True`, the unused parameter will be detected as not belonging to the autograd graph, marked as ready for reduction at the beginning of backward(), and its allreduce bucket slot will be filled with zeros before the allreduce, but variable.grad itself remains undefined. After the allreduce, `!global_unused` means some other processes actually contributed nonzero gradients to that allreduce bucket slot, so we arrive in the above branch and are obligated to both create the grad in this process and populate it with the nonzero values that the allreduce bucket pulled in from the other processes.

(If DDP was constructed with `find_unused_parameters=False`, the case of some parameters that require_grad going unused in some processes is unsupported.  The bucket(s) that are supposed to allreduce those gradients will never get them, the process will never kick off the allreduce for that bucket, causing hangs if other processes did kick off that allreduce's bucket.  I'm curious about details on that subject my own self, see [Slack](https://pytorch.slack.com/archives/CBHSWPNM7/p1590183025075400).)

Reply:
Haha, good corner case. Sounds good !

What about the proposal to use `accumulateGrad` ? Do you think that would be feasible?

Reply:
It's feasible but there's only one branch of accumulateGrad that does what we want, and accumulateGrad is a big function, so I think this makes the code less readable.  As written the intent is pretty obvious, at the cost of stride-logic duplication.  I'd rather keep it as is but I don't feel strongly about it.

Reply:
Sounds good!
Does this needs a comment in accumulate grad to mention that this code should be kept in sync if the AccumulateGrad is ever changed?

Reply:
Ha, in the meantime I decided to compromise on duplication and readability by making a [`clone_obeys_contract`](https://github.com/pytorch/pytorch/pull/34904/files#diff-2fab6123151aca34b081f12ba3d27524R23) helper function.  Both [accumulate_grad.h](https://github.com/pytorch/pytorch/pull/34904/files#diff-ef1a56d24f66b280dcdb401502d6a796R125) and [Reducer.cpp](https://github.com/pytorch/pytorch/pull/34904/files#diff-c269190a925a4b0df49eda8a8f6c5bd3R734) now use the helper function.